### PR TITLE
pose bottomup higherhrnet: model

### PR DESCRIPTION
**Pull request by zhiboniu**
pose bottomup higherhrnet: model

### CODE REVIEW

File: ppdet/modeling/architectures/keypoint_hrhrnet.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
+from __future__ import absolute_import
```
Comment (qingqing01) [inline ppdet/modeling/architectures/keypoint_hrhrnet.py:15]: add copyright

File: ppdet/modeling/architectures/keypoint_hrhrnet.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+                    continue
+                y = max_inds[salvage_joints] // W
+                x = max_inds[salvage_joints] % W
+                offsets = lerp(salvage_joints.nonzero()[0], y, x, heatmap)
+                y = y.astype(np.float32) + offsets[0]
+                x = x.astype(np.float32) + offsets[1]
+                pose_coords[pid][salvage_joints, 0] = y
+                pose_coords[pid][salvage_joints, 1] = x
+                pose_kpts[pid][salvage_joints, 2] = max_scores[salvage_joints]
+        pose_kpts[..., :2] = transpred(pose_coords[..., :2][...,::-1], original_height, original_width, min(H, W))
```
Comment (qingqing01) [inline ppdet/modeling/architectures/keypoint_hrhrnet.py]: transpred 函数是只在模型里用，还是数据预处理也用？

File: ppdet/modeling/architectures/keypoint_hrhrnet.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    __category__ = 'architecture'
+    __shared__ = ['data_format']
+
+    def __init__(self,
+                 backbone='Hrnet',
+                 hrhrnet_head='HigherHrnetHead',
+                 post_process='HrHrnet_PostProcess',
+                 eval_flip = True,
+                 flip_perm = None,
+                 data_format='NCHW'):
```
Comment (qingqing01) [inline ppdet/modeling/architectures/keypoint_hrhrnet.py]: 下面代码看起来 data_format 没用到

File: ppdet/modeling/architectures/keypoint_hrhrnet.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        Args:
+            backbone (nn.Layer): backbone instance
+            hrhrnet_head (nn.Layer): keypoint_head instance
+            bbox_post_process (object): `BBoxPostProcess` instance
+            data_format (str): data format, NCHW or NHWC
+        """
+        super(HigherHrnet, self).__init__(data_format=data_format)
+        self.backbone = backbone
+        self.hrhrnet_head = hrhrnet_head
+        self.post_process = HrHrnet_PostProcess()
```
Comment (qingqing01) [inline ppdet/modeling/architectures/keypoint_hrhrnet.py]: HrHrnetPostProcess

File: ppdet/modeling/architectures/keypoint_hrhrnet.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+@register
+class HigherHrnet(BaseArch):
+    __category__ = 'architecture'
+    __shared__ = ['data_format']
+
+    def __init__(self,
+                 backbone='Hrnet',
+                 hrhrnet_head='HigherHrnetHead',
+                 post_process='HrHrnet_PostProcess',
+                 eval_flip = True,
```
Comment (qingqing01) [inline ppdet/modeling/architectures/keypoint_hrhrnet.py]: 这里加eval参数的话， 模型配置文件里， Train和Eval没办法复用Arch了。

应该是个运行时参数，看起来需要forward时传入吧

File: ppdet/modeling/architectures/keypoint_hrhrnet.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+    def get_pred(self):
+        outputs = {}
+        res_lst, bboxnums = self._forward()
+        outputs['keypoint'] = res_lst
+        outputs['bbox_num'] = bboxnums
+        return outputs
+
+@register
+@serializable
+def HrHrnet_PostProcess(max_num_people=30, heat_thresh=0.2, tag_thresh=1.):
```
Comment (qingqing01) [inline ppdet/modeling/architectures/keypoint_hrhrnet.py]: HrHrnet_PostProcess -> HrHrnetPostProcess

File: ppdet/modeling/architectures/keypoint_hrhrnet.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+                offsets = lerp(salvage_joints.nonzero()[0], y, x, heatmap)
+                y = y.astype(np.float32) + offsets[0]
+                x = x.astype(np.float32) + offsets[1]
+                pose_coords[pid][salvage_joints, 0] = y
+                pose_coords[pid][salvage_joints, 1] = x
+                pose_kpts[pid][salvage_joints, 2] = max_scores[salvage_joints]
+        pose_kpts[..., :2] = transpred(pose_coords[..., :2][...,::-1], original_height, original_width, min(H, W))
+        return pose_kpts, mean_score
+
+    return apply_single
```
Comment (qingqing01) [inline ppdet/modeling/architectures/keypoint_hrhrnet.py]: 可以用class形式写 HrHrnetPostProcess

File: ppdet/modeling/heads/keypoint_hrhrnet_head.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
+import paddle
```
Comment (qingqing01) [inline ppdet/modeling/heads/keypoint_hrhrnet_head.py:15]: add copyright

File: ppdet/modeling/heads/keypoint_hrhrnet_head.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+@register
+class HrHrnetHead(nn.Layer):
+    __inject__ = ['loss']
+
+    def __init__(self,
+                 num_joints,
+                 loss='HrHrnetLoss',
+                 swahr=False,
+                 width=32,
+                 data_format='NCHW'):
```
Comment (qingqing01) [inline ppdet/modeling/heads/keypoint_hrhrnet_head.py]: 同上，目前看 data_format没用到，另外，即使用到，也不用每个接口都设置

File: ppdet/modeling/heads/keypoint_hrhrnet_head.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            has_se=False,
+            freeze_norm=False,
+            name = 'HrHrnetHead_{}'.format(i))
+            for i in range(4)))
+
+        self.interpolate = L.Upsample(2, mode='bilinear')
+        self.concat = L.Concat(dim=1)
+        self._full_name = "HigherHRNet_"+"W"+str(width)
+        if swahr:
+            self._full_name += "_swahr"
```
Comment (qingqing01) [inline ppdet/modeling/heads/keypoint_hrhrnet_head.py]: 这个full_name看起来没用到

File: ppdet/modeling/keypoint_layers.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+def ReLU():
+    return nn.ReLU()
+
+
+def Upsample(scale_factor=None, mode='nearest', align_corners=False):
+    return nn.Upsample(None, scale_factor, mode, align_corners)
+
+
+def MaxPool(kernel_size, stride, padding, ceil_mode=False):
+    return nn.MaxPool2D(kernel_size, stride, padding, ceil_mode=ceil_mode)
```
Comment (qingqing01) [inline ppdet/modeling/keypoint_layers.py]: Why not use nn.Conv2D, nn.Conv2DTranspose,nn.BatchNorm2D nn.ReLU, nn.Upsample , nn.MaxPool2D  in keypoint_hrhrnet_head.py   directly

File: ppdet/modeling/losses/keypoint_loss.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline ppdet/modeling/losses/keypoint_loss.py]: 2021

File: ppdet/modeling/losses/keypoint_loss.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+    def __call__(self, preds, targets):
+        heatmap, mask = targets
+        loss = ((preds - heatmap) ** 2 * mask.cast('float').unsqueeze(1))
+        loss = paddle.clip(loss, min=0, max=2).mean()
+        loss *= self.loss_factor
+        return loss
+
+
+class HeatMapLoss_swahr(object):
```
Comment (qingqing01) [inline ppdet/modeling/losses/keypoint_loss.py]: HeatMapLoss_swahr ->  HeatMapSWAHRLoss ?

File: ppdet/modeling/losses/keypoint_loss.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+import paddle.nn as nn
+import paddle.nn.functional as F
+from ppdet.core.workspace import register, serializable
+
+__all__ = ['HrHrnetLoss']
+
+
+@register
+@serializable
+class HrHrnetLoss(nn.Layer):
```
Comment (qingqing01) [inline ppdet/modeling/losses/keypoint_loss.py:31]: 如果是通用的loss，命名不要带 模型名字

File: ppdet/modeling/architectures/keypoint_hrhrnet.py (hunk 1 above)
Comment (zhiboniu) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py:15]: done

File: ppdet/modeling/architectures/keypoint_hrhrnet.py (hunk 2 above)
Comment (zhiboniu) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py]: 再模型后处理用，但是需要调用预处理里的函数，所以放再operator里了

File: ppdet/modeling/architectures/keypoint_hrhrnet.py (hunk 3 above)
Comment (zhiboniu) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py]: 是的，只是兼容det部分所以留个接口，要删掉吗？

File: ppdet/modeling/architectures/keypoint_hrhrnet.py (hunk 4 above)
Comment (zhiboniu) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py]: done

File: ppdet/modeling/architectures/keypoint_hrhrnet.py (hunk 5 above)
Comment (zhiboniu) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py]: 不影响train和eval的复用。这个参数主要是因为deploy时不能开启flip加的一个标志位（其次时测试选项），开启flip会导致deploy流程出错。因为程序不能区分eval和deploy的区别，需要用户在deploy时手工关掉才行。这个后面需要想个更优雅的解决办法。

File: ppdet/modeling/architectures/keypoint_hrhrnet.py (hunk 6 above)
Comment (zhiboniu) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py]: done

File: ppdet/modeling/heads/keypoint_hrhrnet_head.py (hunk 8 above)
Comment (zhiboniu) [reply ppdet/modeling/heads/keypoint_hrhrnet_head.py:15]: done

File: ppdet/modeling/heads/keypoint_hrhrnet_head.py (hunk 10 above)
Comment (zhiboniu) [reply ppdet/modeling/heads/keypoint_hrhrnet_head.py]: 看起来现在checkpoint保存名称都一样，是否考虑保持名称与模型结构对应起来？

File: ppdet/modeling/losses/keypoint_loss.py (hunk 12 above)
Comment (zhiboniu) [reply ppdet/modeling/losses/keypoint_loss.py]: done

File: ppdet/modeling/losses/keypoint_loss.py (hunk 13 above)
Comment (zhiboniu) [reply ppdet/modeling/losses/keypoint_loss.py]: done

File: ppdet/modeling/losses/keypoint_loss.py (hunk 14 above)
Comment (zhiboniu) [reply ppdet/modeling/losses/keypoint_loss.py:31]: HrHrnet系列专用loss，通用性不强

File: ppdet/modeling/keypoint_layers.py (hunk 11 above)
Comment (zhiboniu) [reply ppdet/modeling/keypoint_layers.py]: 1是直接调用参数有点多看起来不美观。2是配置相同封装一下也方便改，比如初始化Normal(0.001)这种

File: ppdet/modeling/heads/keypoint_hrhrnet_head.py (hunk 9 above)
Comment (zhiboniu) [reply ppdet/modeling/heads/keypoint_hrhrnet_head.py]: 已删

File: ppdet/modeling/architectures/keypoint_hrhrnet.py (hunk 7 above)
Comment (zhiboniu) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py]: done

File: ppdet/modeling/architectures/keypoint_hrhrnet.py (hunk 3 above)
Comment (zhiboniu) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py]: 已删

File: ppdet/modeling/architectures/keypoint_hrhrnet.py (hunk 2 above)
Comment (qingqing01) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py]: 那看下是否在 https://github.com/PaddlePaddle/PaddleDetection/tree/develop/ppdet/modeling 里增加公共函数。

当前 

https://github.com/PaddlePaddle/PaddleDetection/blob/f48cab6332f4f4443f425e9020240f52d2a77675/ppdet/data/transform/batch_operators.py#L30

ppdet/data/transform/batch_operators.py  已经依赖了 ppdet/modeling/


这里，ppdet/modeling又依赖 ppdet/data/transform


会出现 循环引用

File: ppdet/modeling/keypoint_layers.py (hunk 11 above)
Comment (qingqing01) [reply ppdet/modeling/keypoint_layers.py]: 但即使封装，已经有 https://github.com/PaddlePaddle/PaddleDetection/blob/develop/ppdet/modeling/layers.py  文件了

这里封装是通用的，并不是keypoint专用的。

File: ppdet/modeling/architectures/keypoint_hrhrnet.py [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+        return {
+            'backbone': backbone,
+            "hrhrnet_head": hrhrnet_head,
+            "post_process": post_process,
+        }
+
+    def _forward(self):
+        batchsize = self.inputs['image'].shape[0]
+        if self.flip and not self.training and not self.deploy:
+            self.inputs['image'] = paddle.concat((self.inputs['image'], paddle.flip(self.inputs['image'], [3])))
```
Comment (qingqing01) [inline ppdet/modeling/architectures/keypoint_hrhrnet.py]: 一行不超过80个字符

File: ppdet/modeling/backbones/hrnet.py [hunk 16]
```diff
@@ -688,6 +688,7 @@ def __init__(self,
             has_se=self.has_se,
             norm_decay=norm_decay,
             freeze_norm=freeze_norm,
+            multi_scale_output=len(return_idx)>1,
```
Comment (qingqing01) [inline ppdet/modeling/backbones/hrnet.py]: please use pre-commit to format code

可能格式化后是 len(return_idx) > 1

File: ppdet/modeling/architectures/keypoint_hrhrnet.py [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+        right = np.clip(x + 1, 0, W - 1)
+        up = np.clip(y - 1, 0, H - 1)
+        down = np.clip(y + 1, 0, H - 1)
+        offset_y = np.where(
+            heatmap[j, down, x] > heatmap[j, up, x], 0.25, -0.25)
+        offset_x = np.where(
+            heatmap[j, y, right] > heatmap[j, y, left], 0.25, -0.25)
+        return offset_y + 0.5, offset_x + 0.5
+
+    def __call__(self, inputs, original_height, original_width):
```
Comment (qingqing01) [inline ppdet/modeling/architectures/keypoint_hrhrnet.py:143]: 另外，后处理含有很多numpy，现在这样模型导出部署时，后处理咋办？

File: ppdet/modeling/keypoint_layers.py (hunk 11 above)
Comment (zhiboniu) [reply ppdet/modeling/keypoint_layers.py]: 参数初始化会有些不一样，需要验证一下std=0.01是否有影响，higherhrnet系列初始化都用的std=0.001

File: ppdet/modeling/architectures/keypoint_hrhrnet.py (hunk 17 above)
Comment (zhiboniu) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py:143]: 这个后处理太复杂了，导出时不包含这部分。然后python infer里加回来。只是实现c++版可能会比较难

Comment (heavengate) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py:143]: 原作者实现里这块是怎么做的呢，通过numpy做后处理的话后面c++的部署后处理也还是需要解决，毕竟c++部署场景比较多

File: ppdet/modeling/architectures/keypoint_hrhrnet.py (hunk 15 above)
Comment (zhiboniu) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py]: done

File: ppdet/modeling/architectures/keypoint_hrhrnet.py (hunk 17 above)
Comment (zhiboniu) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py:143]: 原作者的实现也是以numpy为主。因为这块他们现在还更多是科研实验方向，比较少考虑工程落地的

File: ppdet/modeling/architectures/keypoint_hrhrnet.py (hunk 2 above)
Comment (zhiboniu) [reply ppdet/modeling/architectures/keypoint_hrhrnet.py]: done, 依赖函数移到keypoints_utils.py

File: ppdet/modeling/heads/keypoint_hrhrnet_head.py (hunk 10 above)
Comment (zhiboniu) [reply ppdet/modeling/heads/keypoint_hrhrnet_head.py]: done,已删

File: ppdet/modeling/keypoint_layers.py (hunk 11 above)
Comment (zhiboniu) [reply ppdet/modeling/keypoint_layers.py]: done，合入layers.py

File: ppdet/modeling/backbones/hrnet.py (hunk 16 above)
Comment (zhiboniu) [reply ppdet/modeling/backbones/hrnet.py]: done，开启pre-commit

**Review by qingqing01 — APPROVED** [review]
